python/lsst/daf/butler/direct_butler/_direct_butler.py

```diff
@@ -1391,11 +1391,19 @@ def removeRuns(self, names: Iterable[str], unstore: bool = True) -> None:
             raise TypeError("Butler is read-only.")
         names = list(names)
         refs: list[DatasetRef] = []
+        all_dataset_types = [dt.name for dt in self._registry.queryDatasetTypes(...)]
         for name in names:
             collectionType = self._registry.getCollectionType(name)
             if collectionType is not CollectionType.RUN:
                 raise TypeError(f"The collection type of '{name}' is {collectionType.name}, not RUN.")
-            refs.extend(self._registry.queryDatasets(..., collections=name, findFirst=True))
+            # Work out the dataset types that are relevant.
+            collections_info = self.collections.x_query_info(name, include_summary=True)
+            filtered_dataset_types = self.collections._filter_dataset_types(
+                all_dataset_types, collections_info
+            )
+            with self._query() as query:
```

Comment:
I have no idea whether it helps for this context to be moved up to the loop over `names`.

Reply:
I don't think that would help, since we're pulling down collection records and summaries by collection already.  There might be some small speedup by putting the `x_query_info` call inside the context manager, if that lets those summaries be shared by both that call and the dataset query.